scripts/cbmc-viewer/block.py

```diff
@@ -0,0 +1,274 @@
+                 reachable=None, srcloc=None):
+        """Initialize block coverage statistics.
+
+        Args:
+
+           xml (str): the name of the xml file containing the block
+           coverage data produced by cbmc.
+
+           srcdir (str): the root of the source directory
+
+           blddir (str): the root of the source directory used to
+           build the goto binary (which may be different from srcdir
+           if the source directory was preprocessed before building)
+
+           reachable (obj): an instance of the Reachable class giving
+           statistics on the statically reachable functions
+        """
+
+        # pylint: disable=too-many-arguments
+
+        self.line = {}
+        self.function = {}
+        self.file = {}
+
+        # Record the list of reachable functions for later use
+        self.reachable = reachable.functions if reachable else []
+        # Record the mapping of filenames computed below for later use
+        # This should be using srcloc.clean_path
+        self.cleanup_filename = Block.cleanup_filename_function(blddir or
+                                                                srcdir)
+
+        if not xml:
+            return
+
+        root = SrcLoc.parse_xml_file(xml, abort=False)
+        if root is None:
+            return
+        srcloc.scan_source_locations(root)
+
+        line_status = Block.line_status(root, srcloc)
+        self.line = Block.line_coverage(line_status)
+        self.function = Block.function_coverage(line_status)
+        self.file = Block.file_coverage(line_status)
+
+    @staticmethod
+    def line_status(root, srcloc):
+        """Compute reachability of a line in a source file."""
+
+        # pylint: disable=too-many-locals
+
+        line_status = {}
+
+        for goal in root.iter("goal"):
+            goal_description = goal.get("description")
+            if not goal_description:
+                break
+            goal_status = goal.get("status")
+            goal_loc = goal.find("location")
+            loc_file = goal_loc.get("file")
+            loc_func = goal_loc.get("function")
+            loc_line = int(goal_loc.get("line"))
+
+            lnst = (linestatus.HIT
+                    if goal_status == "SATISFIED"
+                    else linestatus.MISSED)
+
+            locations = Block.parse_location_description(goal_description)
+            if not locations:
+                locations = [(loc_file, loc_func, loc_line)]
+
+            for loc in locations:
+                src, func, line = loc
+                src = srcloc.clean_path(src)
+
+                if not line_status.get(src):
+                    line_status[src] = {}
+                if not line_status[src].get(func):
+                    line_status[src][func] = {}
+                if not line_status[src][func].get(line):
+                    line_status[src][func][line] = None
+                line_status[src][func][line] \
```

Comment:
While somewhat esoteric (although not at all unheard of), function definitions can be split over multiple files. This however seems to assume a hierarchy where functions are subordinate to files?

Reply:
> I'm curious about this. When would this ever be a good idea?

I've personally used this to make writing an OpenGL wrapper easier; In CBMC we use a similar technique (albeit not for function definitions, but for enums etc) to pregenerate parts of our irep_ids table